### PR TITLE
feat: add message template support for git providers

### LINKED ISSUE #2625: [Bug]: announce: ensure correctness of the GitURL
### What happened?

If the repo is cloned via `SSH`, `.GitURL` will be the SSH URL, which will be consumed by `announce` pipe and send a notification to target. Eventually there will be an SSL URL instead of HTTP URL. [^1] [^2]

[^1]: https://github.com/goreleaser/goreleaser/issues/2622#issuecomment-958358917
[^2]: https://github.com/goreleaser/goreleaser/pull/2623#pullrequestreview-795989193

### How can we reproduce this?

1. Clone the repo using `ssh`
2. Set an announce on the template
3. `.GitURL` eventually will be SSH URL instead of HTTPS

### goreleaser version

47d0e827436f3c2ed8caac9d7f508b2a862c0e00

### GoReleaser Check

- [X] goreleaser check shows no errors

### Search

- [X] I did search for other open and closed issues before opening this.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

Some solutions to tackle this problem:
* We can mention that the `.GitURL` may not be an HTTP URL on the `template.md`. 
* Converting SSH URL to HTTPS URL in the announce stage?
* Separating to `.GitHTTPSURL` and `.GitSSHURL`?

cc: @developer-guy @erkanzileli

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>
Co-authored-by: Furkan Türkal <furkan.turkal@trendyol.com>
Co-authored-by: Erkan Zileli <erkan.zileli@trendyol.com>

Fixes #2625 

to support message templating while announcing for supported git providers such as GitHub, GitLab, Gitea

### CODE REVIEW

**Comment by caarlos0** [comment]
maybe it would be leaner to do put this in a `HTTPGitURL` instead?

**Comment by developer-guy** [comment]
> maybe it would be leaner to do put this in a `HTTPGitURL` instead?

you mean instead of {{ trimsuffix .GitURL ".git" }} ?

**Comment by caarlos0** [comment]
> you mean instead of {{ trimsuffix .GitURL ".git" }} ?

I mean instead of the switch by token type...

**Comment by erkanzileli** [comment]
@caarlos0 can you give an example for us?

**Comment by caarlos0** [comment]
sorry, I thought one thing and typed another completely different...

after taking a better look, here's my idea:
- we already have `client.ReleaseURLTemplate`
  - it is a bit misleading, as its actually the release artifact download template
  - we should probably rename it
- once renamed, we can create a new method called `ReleaseURL` which is in fact the URL to the release
  - this method should return the actual current release url, e.g. `https://github.com/goreleaser/goreleaser/releases/v0.184.0`
- we can then set its value to the context as `ReleaseURL`
- we can also add it as template field
- the templates that are using `.GitURL` can then use `.ReleaseURL` instead

does that makes sense?

**Comment by developer-guy** [comment]
PTAL @caarlos0 🙏

File: internal/pipe/reddit/reddit.go [hunk 1]
```diff
@@ -11,8 +11,9 @@ import (
 )
 
 const (
-	defaultTitleTemplate = `{{ .ProjectName }} {{ .Tag }} is out!`
-	defaultURLTemplate   = `{{ trimsuffix .GitURL ".git" }}/releases/tag/{{ .Tag }}`
+	defaultTitleTemplate     = `{{ .ProjectName }} {{ .Tag }} is out!`
+	defaultGitHubURLTemplate = `{{ trimsuffix .GitURL ".git" }}/releases/tag/{{ .Tag }}`
+	defaultGitLabURLTemplate = `{{ trimsuffix .GitURL ".git" }}/-/releases/{{ .Tag }}`
```
Comment (caarlos0) [inline internal/pipe/reddit/reddit.go:16]: not needed anymore....

File: internal/pipe/reddit/reddit.go [hunk 2]
```diff
@@ -30,8 +31,15 @@ func (Pipe) Default(ctx *context.Context) error {
...
-		ctx.Config.Announce.Reddit.URLTemplate = defaultURLTemplate
+	switch ctx.TokenType {
+	case context.TokenTypeGitHub:
+		ctx.Config.Announce.Reddit.URLTemplate = defaultGitHubURLTemplate
+	case context.TokenTypeGitLab:
+		ctx.Config.Announce.Reddit.URLTemplate = defaultGitLabURLTemplate
+	case context.TokenTypeGitea:
+		ctx.Config.Announce.Reddit.URLTemplate = defaultGitHubURLTemplate
+	default:
+		return fmt.Errorf("invalid client token type: %q", ctx.TokenType)
```
Comment (caarlos0) [inline internal/pipe/reddit/reddit.go:42]: not needed anymore...

```suggestion
	if ctx.Config.Announce.Reddit.URLTemplate == "" {
 		ctx.Config.Announce.Reddit.URLTemplate = defaultURLTemplate
```

File: internal/pipe/reddit/reddit_test.go [hunk 3]
```diff
@@ -14,6 +14,7 @@ func TestStringer(t *testing.T) {
 
 func TestDefault(t *testing.T) {
 	ctx := context.New(config.Project{})
+	ctx.TokenType = context.TokenTypeGitHub
```
Comment (caarlos0) [inline internal/pipe/reddit/reddit_test.go:17]:
```suggestion
```

also not needed

File: internal/pipe/reddit/reddit.go (hunk 1 above)
Comment (caarlos0) [inline internal/pipe/reddit/reddit.go:16]:
```suggestion
	defaultTitleTemplate = `{{ .ProjectName }} {{ .Tag }} is out!`
  	defaultURLTemplate   = `{{ .ReleaseURL }}`
```

File: internal/pipe/twitter/twitter.go [hunk 4]
```diff
@@ -11,7 +11,9 @@ import (
 	"github.com/goreleaser/goreleaser/pkg/context"
 )
 
-const defaultMessageTemplate = `{{ .ProjectName }} {{ .Tag }} is out! Check it out at {{ trimsuffix .GitURL ".git" }}/releases/tag/{{ .Tag }}`
+const (
+	defaultMessageTemplate = `{{ .ProjectName }} {{ .Tag }} is out! Check it out at {{ .ReleaseURL }}`
+)
```
Comment (caarlos0) [inline internal/pipe/twitter/twitter.go:16]:
```suggestion
const defaultMessageTemplate = `{{ .ProjectName }} {{ .Tag }} is out! Check it out at {{ .ReleaseURL }}`
```

File: internal/pipe/telegram/telegram.go [hunk 5]
```diff
@@ -10,7 +10,9 @@ import (
 	"github.com/goreleaser/goreleaser/pkg/context"
 )
 
-const defaultMessageTemplate = `{{ .ProjectName }} {{ .Tag }} is out! Check it out at {{ trimsuffix .GitURL ".git" }}/releases/tag/{{ .Tag }}`
+const (
+	defaultMessageTemplate = `{{ .ProjectName }} {{ .Tag }} is out! Check it out at {{ .ReleaseURL }}`
+)
```
Comment (caarlos0) [inline internal/pipe/telegram/telegram.go:15]:
```suggestion
const defaultMessageTemplate = `{{ .ProjectName }} {{ .Tag }} is out! Check it out at {{ .ReleaseURL }}`
```

**Comment by caarlos0** [comment]
Thanks y'all :)

**Pull request merged**